### ISSUE TITLE
Support FormTypes as shared services

### ISSUE THREAD

**Issue by nsaleh** [symfony/symfony#21447]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2

Hi there,

right now you can register formtypes as services using "form.type" tag. These Services are registered in FormRegistry (using an Extension) and cached there internaly. So if you define a FormType-service not shared, it is still used like a singleton anyways.

My use case is similar to the DoctrineType: 

https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php

DoctrineType has to take care of different instances of "choiceLoaders" and "idreaders". If there was a fresh instance for each time the type is used, this "overhead" could be saved.

As I don't have a deep insight in the form component, there might be side  effects. Before upgrading to 3.2 I just passed a new FormType-Instance , which is now deprecated. That worked very nice.

Thanks,

Nabil

**Comment by HeahDude** [comment]
Hi @nsaleh, thank you for opening this issue.

Actually you're perfectly right, and every form type instance is shared during a request although it can configure many different forms/fields.

>DoctrineType has to take care of different instances of "choiceLoaders" and "idreaders". If there was a fresh instance for each time the type is used, this "overhead" could be saved.

In fact having only one instance of each form types does save some overhead (big time) in the core, and this is why passing other instances has been deprecated in the first place (as of 2.8 using class name).

Instead using some options to change how the form is built dynamically is the way to go. The doctrine type got once the manager registry and can be used to configure many fields with different `class`, `em` or `query_builder` options (see https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php#L147-L151)  because each time `buildForm` is called it gets a specific instance of builder (the form being built) with the options set for that one only.

You can also have many types to extend a particular one using it as parent to have more specific configuration, this is what happens for the `ChoiceType`, it has many children like the `DoctrineType`, `CountryType`, `LocaleType`...

In conclusion, this is not something that should change IMHO.

**Comment by xabbuh** [comment]
I am closing here as this is indeed the intended behaviour and @HeahDude described some ideas about how you could solve your problem. Thank you for understanding.

**Comment by nsaleh** [comment]
Okay If singleton types are the big winners in performance, I'll go with it, thanks for the profound answer!
